### ISSUE TITLE
Strange behavior of the join() function with arrays containing empty strings

### ISSUE THREAD

**Issue by iomicifikko** [OpenRefine/OpenRefine#3290]
### To Reproduce
1. First, click on a column menu
2. Then, select "Edit cells" > "Transform"
3. Finally, use this expression: `["",1,2].join("|")` and press OK button.

### Current Results

The string **1|2**

### Expected Behavior

The string **|1|2**

### Screenshots

![image](https://user-images.githubusercontent.com/61201406/97579979-05ccfe80-19f3-11eb-9cdf-d67c116c282f.png)

### Versions
 - Operating System: Windows 10 64bit 
 - Browser Version: Chrome 86.0.4240.111 (official build) (64 bit) 
 - JRE or JDK Version: Embedded JAVA with OpenRefine 3.4.1 package for Windows (but i tested the same behavior on other OpenRefine/Java versions 
 - OpenRefine: 3.4.1, 3.3 

### Additional context

The same problem happens also if the last item of the array is an empty string.

**Comment by thadguidry** [comment]
I strongly feel this should be another parameter option to join().  I like the current convenience of collapsing the empty elements, but I can see it might be inconvenient for rows of data where you want to say "no, don't collapse empty elements during my join's"?  
Maybe it looks like `join(array a, string sep, optional boolean RemoveEmptyElements)`?

**Comment by iomicifikko** [comment]
@thadguidry 
I think that your suggestion is an optimal solution because gives the choice to the user, who always knows what to expect.
For knowledge of all: the join function axctually remove any **blank**/**null** value inside the array.

**Comment by wetneb** [comment]
I agree this behaviour is curious. The problem now is, as always, how to fix this without breaking existing workflows which rely on this peculiar behaviour.

**Comment by tfmorris** [comment]
This behavior is the reason I added these TODOs a few weeks ago:

https://github.com/OpenRefine/OpenRefine/blob/82b1cf0bfaf3c8ab57a12e0f2004bdeb71432383/main/tests/server/src/com/google/refine/expr/functions/arrays/JoinTests.java#L47-L54

This is just one instance of many "special" cases for nulls and empty strings which are very irregular.

**Comment by singhakshita** [comment]
can i work on this issue ?   
also as @thadguidry suggested should we extend the function to  have one more boolean parameter of knowing  whether to  join the empty the string or not ,into all array.join methods ?
what's the final decision  of expected behaviour taken on this issue?

**Comment by tfmorris** [comment]
I think @thadguidry 's proposal is a reasonable approach to fixing this while preserving backward compatibility, ie extend to the function signature to 
```
join(array a, string sep, optional boolean preserveEmpty default=false)
```

but the original report seems to indicate that behavior changes depending on the position of the empty string in the array and the tests indicate that behavior is different between NULLs and empty strings, when we normally treat those two categories the same.

**Comment by tfmorris** [comment]
I've looked into this and I think it's just a plain bug which doesn't deserve compatibility protection. Currently all leading empty strings are ignored due to a bug in the algorithm for deciding if we've already added an element (it uses the length, which is 0 for empty strings).

Current
```
['',2,3].join('|') =>  "2|3"
['','',''].join('|') => ""
```
Proposed
```
['',2,3].join('|') =>  "|2|3"
['','',''].join('|') => "||"
```

> The same problem happens also if the last item of the array is an empty string.

I can't reproduce this part of the bug report. 

```
[1,2,''].join('|') =>  "1|2|"
```

Note that I'm intentionally ignoring any discussion of `null`s which is a separate topic.
